### PR TITLE
Add telemetry to api server

### PR DESCRIPTION
**Pull request by keathley**
This PR adds telemetry spans to the api server. My hypothesis is that the API Server is a bottleneck on the system and adding substantial overhead under high load scenarios. This telemetry should help to prove or disprove my hypothesis.

### CODE REVIEW

File: lib/spandex_datadog/api_server.ex [hunk 1]
```diff
@@ -144,47 +147,51 @@ defmodule SpandexDatadog.ApiServer do
           agent_pid: agent_pid
         } = state
       ) do
-    all_traces = [trace | waiting_traces]
+    :telemetry.span([:spandex_datadog, :handle_call, :send_trace], %{trace: trace, state: state}, fn ->
```
Comment (GregMefford) [inline lib/spandex_datadog/api_server.ex:150]: I need to give this some more thought, but one initial thing here is that I don't love the name of this event because I feel like the `:handle_call` is an internal detail that should be replaced with a more-permanent name, but I'm not sure what to call it at the moment. `:send_trace` above makes sense because that's the public API that's being called to "send" the trace from your app code into Spandex, but I feel like this needs some "official" name like `:forward_trace` for actually sending the trace from Spandex's buffer to Datadog. 🤔

Comment (GregMefford) [reply lib/spandex_datadog/api_server.ex:150]: I'm also struggling with the thought that one of these events is fired from the caller's process and the other fires in the GenServer process. That's probably not bad but I'm trying to think through the ramifications.

Comment (keathley) [reply lib/spandex_datadog/api_server.ex:150]: If it helps this PR get merged I can remove the telemetry spans from inside the GenServer. I only *really* need to capture the client telemetry.

Comment (GregMefford) [reply lib/spandex_datadog/api_server.ex:150]: I started working on a branch to re-arrange things a little so that it would make sense to have telemetry for the client side like you mentioned, but also telemetry from inside the GenServer for enqueuing a trace and for flushing a batch of traces. We should be able to get it all sorted out today/tomorrow I think.

Comment (keathley) [reply lib/spandex_datadog/api_server.ex:150]: Honestly, if it turns out that the client side of things is becoming a bottleneck, I have a few improvements for this flow I've already been working on. I want to prove that this is an issue before I submit them as a PR though.

Comment (GregMefford) [reply lib/spandex_datadog/api_server.ex:150]: Yeah, I think we should go ahead and ship some minimal telemetry on the client side then, and we can separately clean up/optimize/add more telemetry afterword.

The only think holding up the basic telemetry around the client call is some docs about it, which I can help with. Primarily, I want to make sure we document the fact that this `:telemetry.span` would fire after it's too late to update the current trace, because the trace has already been sent to the APIServer, so you won't be able to include this time as a span in the overall trace unless you `continue_trace` and send a separate trace with only this span in it. If you did _that_ though, you'd have to be careful to not set `batch_size` to `1` or else it will infinitely send new traces of it sending traces.

File: lib/spandex_datadog/api_server.ex [hunk 2]
```diff
@@ -106,8 +106,11 @@ defmodule SpandexDatadog.ApiServer do
   """
   @spec send_trace(Trace.t(), Keyword.t()) :: :ok
   def send_trace(%Trace{} = trace, opts \\ []) do
-    timeout = Keyword.get(opts, :timeout, 30_000)
-    GenServer.call(__MODULE__, {:send_trace, trace}, timeout)
+    :telemetry.span([:spandex_datadog, :send_trace], %{trace: trace}, fn ->
+      timeout = Keyword.get(opts, :timeout, 30_000)
```
Comment (GregMefford) [inline lib/spandex_datadog/api_server.ex:110]: I think in order for this option to be accepted, we'll need to add it to the Optimal schema above.
We [will probably remove Optimal as a dependency](https://github.com/spandex-project/spandex/pull/124) because it has some performance issues, but that'll be a separate PR.

Comment (GregMefford) [reply lib/spandex_datadog/api_server.ex:110]: Oh wait, I misread that - this is the `:timeout` option to `APIServer.send_trace` and it was already there before. Never mind! 😅

File: lib/spandex_datadog/api_server.ex [hunk 3]
```diff
@@ -84,7 +84,7 @@ defmodule SpandexDatadog.ApiServer do
   """
   @spec init(opts :: Keyword.t()) :: {:ok, State.t()}
   def init(opts) do
-    {:ok, agent_pid} = Agent.start_link(fn -> 0 end, name: :spandex_currently_send_count)
+    {:ok, agent_pid} = Agent.start_link(fn -> 0 end)
```
Comment (GregMefford) [inline lib/spandex_datadog/api_server.ex:87]: @zachdaniel Do you know/recall whether this is actually needed for anything? I was struggling to figure out how to start up this GenServer in tests because it wants to register this Agent by name, and it can't because it's already registered in the `setup_all` block. As far as I can tell, this isn't even needed and it works just fine without it because we track the PID in the GenServer state anyway.

File: lib/spandex_datadog/api_server.ex (hunk 1 above)
Comment (GregMefford) [reply lib/spandex_datadog/api_server.ex]: Ok @keathley, I removed the telemetry from inside the GenServer and added some docs / tests.
I believe this still does what you intended, but please let me know if I missed something!

File: lib/spandex_datadog/api_server.ex (hunk 3 above)
Comment (zachdaniel) [reply lib/spandex_datadog/api_server.ex:87]: Nothing comes to mind. If it works without it then I say leave it out.

**Comment by GregMefford** [comment]
Hm, working on getting it to reliably crash and fire the `:exception` event in tests. 😫

**Review by zachdaniel — APPROVED** [review]
